### ISSUE TITLE
Add .Expand() to resolve $ref's

### ISSUE THREAD

**Issue by fmvilas** [xeipuuv/gojsonschema#156]
Hi, I'm working on a code generator for the [AsyncAPI specification](https://asyncapi.org). It is very similar to OpenAPI. I wanted to use your package but I need to expand all the `$ref` pointers and get a proper JSON object. Is it something that you would like to add? Maybe it already exists and I'm missing something.

I can take care of it if you want.

Thanks!

**Comment by handrews** [comment]
[disclaimer: I'm not associated with this project, but I am part of the JSON Schema spec team]

It's not generally possible to expand all `$ref` pointers and produce valid JSON, as you may end up with cycles in your dereferenced data structure.  The meta-schemas are an example of this, as they use `{"$ref": "#"}` extensively to build a tree structure (many values within a JSON Schema are themselves JSON Schemas).

**Comment by fmvilas** [comment]
Hey @handrews, thanks for your response. I know there are certain cases where you can't perform the "dereferencing" but there are lots of cases where it's actually useful and possible. Usually, OpenAPI and AsyncAPI documents don't have circular references nor self-references. They are very simple JSON schema documents.

So from my perspective, this is something that's useful even with its limitations.

An example of a package doing it in Node.js: https://github.com/BigstickCarpet/json-schema-ref-parser/blob/master/docs/ref-parser.md#dereferenceschema-options-callback

**Comment by handrews** [comment]
Yes, it definitely has some uses (I'm familiar with that package, actually).  But I think it is important to advertise feature correctly including the limitations.  There are very simple structures that can't be implemented without circular references.

Personally, I find API description formats that don't allow structures like trees to be described to be non-starters, but I realize many APIs can get away without it.

**Comment by fmvilas** [comment]
> But I think it is important to advertise feature correctly including the limitations.

Totally agree.

**Comment by johandorland** [comment]
What do you expect `Expand()` to return as a type? The Node.js package you mentioned seems really cool, but I haven't figured out how something similar might look in Go.
